'Introducing Bloom Filters for Valkey' Blog (#229)
valkey-io/valkey-io.github.io: content/blog/2025-04-02-introducing-bloom-filters.md

+++
title= "Introducing Bloom Filters for Valkey"
description = "Learn how to use bloom filters to perform large-scale membership testing with significant memory savings."
date= 2025-04-02 01:01:01
authors= [ "karthiksubbarao"]
+++

We are introducing Bloom Filters as a new data type in Valkey via [valkey-bloom](https://github.com/valkey-io/valkey-bloom/), an official Valkey Module which is compatible with Valkey versions >= 8.0. Bloom filters provide efficient, large-scale membership testing, improving performance and offering significant memory savings for high-volume applications.

> As an example, previously, Valkey users could use the `SET` data type for handling advertisement deduplication workloads. However, since every item in the set is allocated, the problem with this approach is high memory usage. This article demonstrates how using the bloom filter data type from valkey-bloom achieves >93% memory savings for the same workload, while exploring its implementation, technical details, and practical recommendations.

Reviewer: Reflow the lines at some point, it makes it much easier to review and leave comments. I like doing it after each sentence, but you can also do it at line lengths.

Reviewer: On this line, the use case isn't described very clearly before explaining it uses a lot of memory. Ideally we would fully describe the problem up front, describe the solution, then talk about how it's implemented.

Author: > Reflow the lines at some point, it makes it much easier to review and leave comments. I like doing it after each sentence, but you can also do it at line lengths.

Sure, I can move every sentence into a new line

Author: > On this line, the use case isn't described very clearly before explaining it uses a lot of memory. Ideally we would fully describe the problem up front, describe the solution, then talk about how it's implemented.

Thank you - I added more context in the intro now. Let me know if I should add anything more.

## Introduction

Bloom filters are a space efficient probabilistic data structure that allows adding elements and checking whether elements exist. False positives are possible where a filter incorrectly indicates that an element exists, even though it was not added. However, Bloom Filters guarantee that false negatives do not occur, meaning it can never be the case that an element was added successfully and a filter reports it as not existing. Bloom Filters were first introduced in a paper from 1970 by Burton H. Bloom.

> <img src="/assets/media/pictures/bloomfilter_bitvector.png" style="width: 75%;" alt="Bloom Filter Bit Vector">

Reviewer: Should be in markdown, we don't want to directly embed the html tags.

Author: Updated

> <p style="text-align: center;"><i>Image taken from <a href="https://upload.wikimedia.org/wikipedia/commons/a/ac/Bloom_filter.svg">source</a></i></p>

Reviewer: Let's pull the image and credit it or make our own version.

Author: Currently, I am crediting this by saying "Image taken from source" where source links to the wikipedia source.

Is there anything additional to do to credit it here?

<img width="507" alt="Screenshot 2025-04-03 at 10 24 06 AM" src="https://github.com/user-attachments/assets/6ae9101a-f548-4cde-8228-c4f77e4d0a69" />

Without bloom filters, applications could use the `SET` Valkey data type such that they have a unique `SET` for every advertisement. Then, they can use the `SADD` command track every customer who has already seen this particular advertisement by adding them to the set. This means we have 500 sets, each with 5M members. This will require ~152.57 GB of `used_memory` on a Valkey 8 server.

With bloom filters, applications can create a unique bloom filter for every advertisement with the `BF.RESERVE` or `BF.INSERT` command. Here, they can specify the exact capacity they require: 5M - which means 5M items can be added to the bloom filter. For every customer that the advertisement is shown to, the application can add the UUID of the customer onto the specific filter. So, we have 500 bloom filters, each with a capacity of 5M. This will require variable memory depending on the false positive rate. In all cases (even stricter false positive rates), we can see there is a significant memory optimization compared to using the `SET` data type.

> <table width="100%" border="1" style="border-collapse: collapse; border: 1px solid black" cellpadding="8">

Reviewer: Same, also make this a markdown table.

Author: Just wanted to understand the reasoning behind keeping blogs to strictly be markdown context.

It makes sense that for documentation we do not use HTML since it does not get rendered clearly in the `man`. But blogs do not get listed in the `man`.

Is the reason for not including HTML in blogs that we can go wrong in rendering this on different platforms? Example: Platform does not support HTML or HTML code is not responsive and does not scale based on the screen size.

The HTML code looks better for tables since it includes a border, and we can add colors to the columns to highlight details

#### Pure Markdown Table:

<img width="685" alt="Screenshot 2025-04-03 at 10 16 27 AM" src="https://github.com/user-attachments/assets/39fbd134-3c66-4874-ac2c-518c8fbf821e" />

#### HTML Table:
<img width="656" alt="Screenshot 2025-04-03 at 10 16 32 AM" src="https://github.com/user-attachments/assets/d7ebdaa8-5a68-442c-bac5-f0e87f2efa5d" />

<tr>
<th>Number of Bloom Filters</th>
<th>Capacity</th>
<th>FP Rate</th>
<th>FP Rate Description</th>
<th>Total Used Memory (GB)</th>
<th>Memory Saved % compared to SETS</th>
</tr>
<tr>
<td>500</td>
<td>5000000</td>
<td>0.01</td>
<td>One in every 100</td>
<td>2.9</td>
<td style="background-color: #90EE90;">98.08%</td>
</tr>
<tr>
<td>500</td>
<td>5000000</td>
<td>0.001</td>
<td>One in every 1K</td>
<td>4.9</td>
<td style="background-color: #90EE90;">96.80%</td>
</tr>
<tr>
<td>500</td>
<td>5000000</td>
<td>0.00001</td>
<td>One in every 100K</td>
<td>7.8</td>
<td style="background-color: #90EE90;">94.88%</td>
</tr>
<tr>
<td>500</td>
<td>5000000</td>
<td>0.0000002</td>
<td>One in every 5M</td>
<td>9.8</td>
<td style="background-color: #90EE90;">93.60%</td>
</tr>
</table>
<br>
<div style="display: flex; align-items: center;">
<p style="flex: 1;">In this example, we are able to benefit from 93% - 98% savings in memory usage when using Bloom Filters compared to the SET data type. Depending on your workload, you can expect similar results.</p>
<img src="/assets/media/pictures/bloomfilter_memusage.png" style="width: 40%;" alt="SET vs Bloom Filter Memory Usage Comparison">
</div>

> ## Limits and recommendations

Author: ```suggestion
## Large Bloom Filter and Recommendations
```

> To ensure optimal server performance during serialization and deserialization of bloom filters, we have added a limit on their memory usage per object. The default memory usage limit of a bloom filter is defined by the `BF.BLOOM-MEMORY-USAGE-LIMIT` module configuration which has a default value of 128 MB. However, the limit can be tuned using the configuration above.

Author: ```suggestion
To ensure optimal server performance during serialization and deserialization of bloom filters, we have added validation on the memory usage per object. The default memory usage limit of a bloom filter is defined by the `BF.BLOOM-MEMORY-USAGE-LIMIT` module configuration which has a default value of 128 MB. However, the value can be tuned using the configuration above.
```

<table width="100%" border="1" style="border-collapse: collapse; border: 1px solid black" cellpadding="8">
<tr>
<th>Non Scaling Filter - Capacity</th>
<th>FP Rate</th>
<th>Memory Usage (MB)</th>
<th>Notes</th>
</tr>
<tr>
<td>112M</td>
<td>0.01</td>
<td>~128</td>
<td>Default FP Rate and Default Memory Limit</td>
</tr>
<tr>
<td>74M</td>
<td>0.001</td>
<td>~128</td>
<td>Custom FP Rate and Default Memory Limit</td>
</tr>
<tr>
<td>448M</td>
<td>0.01</td>
<td>~512</td>
<td>Default FP Rate and Custom Memory Limit</td>
</tr>
<tr>
<td>298M</td>
<td>0.001</td>
<td>~512</td>
<td>Custom FP Rate and Custom Memory Limit</td>
</tr>
</table>

## Performance 

The bloom commands which involve adding items or checking the existence of items have a time complexity of O(N * K) where N is the number of hash functions used by the bloom filter and K is the number of elements being inserted. This means that both `BF.ADD` and `BF.EXISTS` are both O(N) as they only operate on one item.

> In case of scalable bloom filters, with every scale out, we increase the number of checks (using hash functions of each sub filter) performed during any add / exists operation. For this reason, it is recommended that users choose a capacity and expansion rate after evaluating the use case / workload to avoid several scale outs and reduce the number of checks.

Reviewer: How does a user choose the right values? This isn't very prescriptive for the average user.

Author: I made changes here: https://github.com/valkey-io/valkey-io.github.io/pull/229/commits/29a1e1b1d7e022b9cf2b4d8f29fb9f35d7058b92

The other bloom filter commands are O(1) time complexity: `BF.CARD`, `BF.INFO`, `BF.RESERVE`, and `BF.INSERT` (when no items are provided).

## Conclusion

valkey-bloom offers an efficient solution for high-volume membership testing through bloom filters, providing significant memory usage savings compared to traditional data types. This enhances Valkey's capability to handle various workloads including large-scale advertisement / event deduplication, fraud detection, and reducing disk / backend lookups more efficiently.

> To learn more about [valkey-bloom](https://github.com/valkey-io/valkey-bloom/), you can read about the data type [here](https://valkey.io/topics/bloomfilters/) and follow the [quick start guide](https://github.com/valkey-io/valkey-bloom/blob/unstable/QUICK_START.md) to try it yourself. Additionally, to use valkey-bloom on Docker (along with other official modules), you can check out the [Valkey Extensions Docker Image](https://hub.docker.com/r/valkey/valkey-extension).

Author: ```suggestion
To learn more about [valkey-bloom](https://github.com/valkey-io/valkey-bloom/), you can read about the data type [here](https://valkey.io/topics/bloomfilters/) and follow the [quick start guide](https://github.com/valkey-io/valkey-bloom/blob/1.0.0/QUICK_START.md) to try it yourself. Additionally, to use valkey-bloom on Docker (along with other official modules), you can check out the [Valkey Extensions Docker Image](https://hub.docker.com/r/valkey/valkey-extension).
```
